dashboard/utils.py

```diff
@@ -303,6 +303,21 @@ def get_all_final_grades(self):
                 final_grades[course_id] = final_grade
         return final_grades
 
+    def get_all_enrolled_course_runs(self):
+        """
+        Returns a list of course_runs fro which the user is either enrolled
+        and has a final grade
+
+        Returns:
+            list: list of course_runs
+        """
+        course_runs = []
+        final_grades = self.get_all_final_grades()
+        for course_id in self.course_ids:
+            if course_id in final_grades or self.enrollments.is_enrolled_in(course_id):
+                course_runs.append(CourseRun.objects.get(edx_course_key=course_id))
+        return course_runs
```

Comment:
Can you change this logic to make a list of course ids, then use `CourseRun.objects.filter(edx_course_key__in=course_ids)` here? It would reduce the number of database queries which would go a tiny bit toward reducing the [n+1 selects](http://stackoverflow.com/questions/97197/what-is-select-n1) problem we have (though since we are still doing this for every user it won't make a big difference for `recreate_index` here)